data_safe_haven/infrastructure/project_manager.py

```diff
@@ -72,10 +72,19 @@ def __init__(
     @property
     def pulumi_extra_args(self) -> dict[str, Any]:
         extra_args: dict[str, Any] = {}
-        if self.logger.isEnabledFor(logging.DEBUG):
+        # Produce verbose Pulumi output if running in verbose mode
+        if self.logger.console_handler.level <= logging.DEBUG:  # type: ignore [attr-defined]
```

Comment:
I don't love the way that we added the `console_handler` property during the construction of the logging object. It means that we get these type errors each time we use it. Is there any other way around this (I know inheriting from `logging.Logger` is a bit of a mess).

Reply:
Me too.

When making the logging changes, I found that creating a child of `logger.Logger` was a pain.
There is a function to set what class `logging.getLogger()` returns, but it applies globally! _I.e._ all other loggers created by other libraries would be of that class.
Feels very much like you are fighting against the logging module.

Can you see a better solution?

Reply:
Possibly a function in `logging/logger.py` like this?

```
def console_handler() -> RichHandler:
  return next([h for h in get_logger().handlers if isinstance(h, RichHandler)])
```

Reply:
Yes that's not too bad.

I did think about doing something like that but it seemed like a waste to run that code every time.
That said, you don't need to do it very often and it probably doesn't take long anyway.